python_shader/_generator_base.py

```diff
@@ -1,10 +1,18 @@
 """
 Implements the base class for generating SpirV code.
+
+Notes on exceptions for code that parses/converts/compiles shading code:
+
+* Raise ShaderError when the user-provided code cannot be compiled.
+* Raise RuntimeError when the parser/compiler is in an unexpected state.
+* Raise NotImplementedError only to indicate that it's an abstract method.
+* Always provide an error message.
```

Comment:
@Korijn If you could have a look at these 4 lines. The rest is just implementing that :)

Reply:
Ok!

I think shadererror could also be syntaxerror or valueerror which are both builtins :) you decide!

Otherwise LGTM

Reply:
(value error indicates a user provided value is invalid, so that's how I could see that being appropriate here)

Reply:
I do see the value in having an instantly recognizable error type though! :)

Reply:
Yeah, I considered those (and other builtins). But for a `ValueError` I think of an int being out of bounds or something. Shaders feel to be more than just "a value" :)  

`SyntaxError` is not always appropriate, because errors like these can rarely be called `SyntaxErrors`.